From 8395408a9a6372245cac329d98d3cc6a932096fe Mon Sep 17 00:00:00 2001
From: Maksim Sokolskiy <misokolsky@gmail.com>
Date: Mon, 17 Oct 2016 19:25:04 +0300
Subject: [PATCH] feat(gamafication_metric): add `d3.js animated pie chart` for
 user points

Changes:
- add js, css directly into dashboard.html
- change `GAMMA_PROPERTIES` structure: `API_URL` should contains basic API url w/o `gamma-profile`
- API request from browser uses `GAMMA_PROPERTIES`

TODO:
- move js, css into includable files

feat(dashboard): add animated rind to render user rewards by category

makrup(dashboard): fix markup for gamification widgets

markup(dashboard.html): add combined category/points half/donut `highcharts ` widget, progress, accumulative progress widgets, badges

markup(dashboard): add second YAxis for accumulative graph
---
 common/djangoapps/gamification_metric/api_calls.py |   7 +-
 common/djangoapps/gamification_metric/tasks.py     |   4 +-
 common/djangoapps/student/views.py                 |   8 +
 lms/templates/courseware/progress.html             |  27 ++-
 lms/templates/dashboard.html                       | 185 ++++++++++++++++++++-
 5 files changed, 220 insertions(+), 11 deletions(-)

diff --git a/common/djangoapps/gamification_metric/api_calls.py b/common/djangoapps/gamification_metric/api_calls.py
index f577a3e..9132eac 100644
--- a/common/djangoapps/gamification_metric/api_calls.py
+++ b/common/djangoapps/gamification_metric/api_calls.py
@@ -38,4 +38,9 @@ class APICalls(object):
             'App-key': self.GAMMA_PROPERTIES['APP_KEY'],
             'App-secret': self.GAMMA_PROPERTIES['APP_SECRET']
         }
-        requests.put(self.GAMMA_PROPERTIES['API_URL'], data=data, headers=headers, verify=False)
+        requests.put(
+            self.GAMMA_PROPERTIES['API_URL']+'gamma-profile/',
+            data=data,
+            headers=headers,
+            verify=False
+        )
diff --git a/common/djangoapps/gamification_metric/tasks.py b/common/djangoapps/gamification_metric/tasks.py
index 376b570..71d29aa 100644
--- a/common/djangoapps/gamification_metric/tasks.py
+++ b/common/djangoapps/gamification_metric/tasks.py
@@ -5,4 +5,6 @@ from api_calls import APICalls
 
 @task()
 def send_api_request(data):
-    APICalls().api_call(**data)
+    api = APICalls()
+    if api.is_enabled:
+        api.api_call(**data)
diff --git a/common/djangoapps/student/views.py b/common/djangoapps/student/views.py
index 97823cc..3d2282e 100644
--- a/common/djangoapps/student/views.py
+++ b/common/djangoapps/student/views.py
@@ -709,6 +709,13 @@ def dashboard(request):
 	else:
 	    course_enrollments_progress.append(enrollment)
 
+    if (
+        settings.FEATURES.get('ENABLE_GAMMA', False) and
+        settings.FEATURES.get('GAMMA_PROPERTIES', {}) and
+        settings.FEATURES.get('GAMMA_PROPERTIES', {}).get('API_URL')
+    ):
+        api_url = settings.FEATURES.get('GAMMA_PROPERTIES', {}).get('API_URL')
+
     context = {
         'enrollment_message': enrollment_message,
         'redirect_message': redirect_message,
@@ -743,6 +750,7 @@ def dashboard(request):
         'nav_hidden': True,
         'course_programs': course_programs,
     	'honor_code_accepted': is_honor_code_accepted(request),
+        'api_url': api_url,
     }
 
     return render_to_response('dashboard.html', context)
diff --git a/lms/templates/courseware/progress.html b/lms/templates/courseware/progress.html
index 26a2e2a..92df31f 100644
--- a/lms/templates/courseware/progress.html
+++ b/lms/templates/courseware/progress.html
@@ -29,6 +29,19 @@ from django.utils.http import urlquote_plus
 <script>
   ${progress_graph.body(grade_summary, course.grade_cutoffs, "grade-detail-graph", not course.no_grade, not course.no_grade) | h}
 </script>
+<script type="text/javascript" charset="utf-8">
+$(document).ready(function () {
+    var domain_address = document.location.origin
+    $(".course-referral-button").click(function () {
+        $.post(domain_address + "/api/get_referral_hash_key/", {'course_id': '${str(course.id)}'},
+            function (data) {
+                $('.course-referral-link').val(domain_address + data.hashkey).show();
+                document.execCommand("copy");
+            });
+
+    });
+});
+</script>
 </%block>
 
 
@@ -136,8 +149,8 @@ padding: 14px 0px 14px;
 <div class="progress-info-block">
 	<span class='icon-info'>i</span>
 	<div class="text-block">Your progress will be depicted in the graph once you start taking tests.</div>
-</div> 
-        % elif answered == 0 and reset is not None : 
+</div>
+        % elif answered == 0 and reset is not None :
 <div class="progress-info-block">
         <span class='icon-info'>i</span>
         <div class="text-block">The course has been reset successfully. Please take the course again, to get the fresh scores.</div>
@@ -183,11 +196,15 @@ $("#reset_button").click(function(){
 			}
                 }
         });
-}); 
+});
 </script>
 
         % endif
       % endif
+        <div class="course-referral">
+            <button class="course-referral-button">Invite friend</button>
+            <input class="course-referral-link" type="text" size="70" disabled>
+        </div>
       % if staff_access and studio_url is not None:
         <div class="wrap-instructor-info">
           <a class="instructor-info-action studio-view" href="${studio_url | h}">${_("View Grading in studio")}</a>
@@ -458,7 +475,7 @@ $("#reset_button").click(function(){
 </div>
 <script type="text/javascript">
 $(document).ready(function(){
-	setTimeout(function(){	
+	setTimeout(function(){
 		var index = 0;
                 $(".scores h3").each(function(){
                         if(!$(this).hasClass("no-scores")){
@@ -469,5 +486,3 @@ $(document).ready(function(){
 	}, 2000);
 });
 </script>
-
-
diff --git a/lms/templates/dashboard.html b/lms/templates/dashboard.html
index 63ef8ca..935ab95 100644
--- a/lms/templates/dashboard.html
+++ b/lms/templates/dashboard.html
@@ -79,8 +79,178 @@ if(localStorage["honor_code_skipped"] == undefined || localStorage["honor_code_s
         $("input[name='enrollment_action'", $("#class_enroll_form")).val("enroll")
         $("#class_enroll_form").submit();
       }
+
     });
   </script>
+
+<script>
+$(function() {
+    $.get(
+        "${api_url}"+"points/",
+        data={"username": "${ user.username | h }"},
+        function( points ) {
+            $.get(
+                "${api_url}"+"chart/",
+                data={"username": "${ user.username | h }"},
+                function( data ) {
+                    var events = [];
+                    for (var key in data) {
+                        events.push({name: key, y : data[key]});
+                    }
+                    $('#container-chart').highcharts({
+                        chart: {
+                            plotBackgroundColor: null,
+                            plotBorderWidth: 0,
+                            plotShadow: false
+                        },
+                        title: {
+                            text: '<h1><b>'+points.points+'</b></h1>',
+                            align: 'center',
+                            verticalAlign: 'middle',
+                            y: 65
+                        },
+                        tooltip: {
+                            pointFormat: '{series.name}: <b>{point.percentage:.1f}%</b>'
+                        },
+                        plotOptions: {
+                            pie: {
+                                dataLabels: {
+                                    enabled: true,
+                                    distance: -50,
+                                    style: {
+                                        fontWeight: 'bold',
+                                        color: 'white'
+                                    }
+                                },
+                                startAngle: -90,
+                                endAngle: 90,
+                                center: ['50%', '75%']
+                            }
+                        },
+                        series: [{
+                            type: 'pie',
+                            name: 'Browser share',
+                            innerSize: '50%',
+                            data: events
+                        }]
+                    });
+                }
+            );
+        }
+    );
+});
+</script>
+
+<script>
+$(function () {
+    $.get(
+        "${api_url}"+"progress/",
+        data={"username": "${ user.username | h }"},
+        function( data ) {
+            var points_by_day = [];
+            var accumulative_data = [];
+            for (var i in data) {
+                points_by_day.push([Date.parse(data[i].date), data[i].points]);
+                var accum = 0;
+                if (accumulative_data.length != 0) {
+                    accum = accumulative_data[accumulative_data.length-1][1] + data[i].points;
+                }
+                else {
+                    accum = data[i].points;
+                }
+                accumulative_data.push([Date.parse(data[i].date), accum]);
+            }
+
+            $('#container-progress').highcharts({
+                chart: {
+                    type: 'spline'
+                },
+                title: {
+                    text: 'Student progress'
+                },
+                subtitle: {
+                    text: 'Shows Student points by day'
+                },
+                xAxis: {
+                    type: 'datetime',
+                    dateTimeLabelFormats: {
+                        month: '%e. %b',
+                        year: '%b'
+                    },
+                    title: {
+                        text: 'Date'
+                    }
+                },
+                yAxis: [{
+                    title: {
+                        text: null
+                    },
+                    labels: {
+                        align: 'left',
+                        x: 3,
+                        y: 16,
+                        format: '{value:.,0f}'
+                    },
+                    showFirstLabel: false
+                }, {
+                    gridLineWidth: 0,
+                    opposite: true,
+                    title: {
+                        text: null
+                    },
+                    labels: {
+                        align: 'right',
+                        x: -3,
+                        y: 16,
+                        format: '{value:.,0f}'
+                    },
+                    showFirstLabel: false
+                }],
+                tooltip: {
+                    headerFormat: '<b>{series.name}</b><br>',
+                    pointFormat: '{point.x:%e. %b}: {point.y:.2f}'
+                },
+
+                plotOptions: {
+                    spline: {
+                        marker: {
+                            enabled: true
+                        }
+                    }
+                },
+
+                series: [{
+                    name: 'Student progress',
+                    type: 'column',
+                    data: points_by_day
+                },
+                {
+                    name: 'Accumulative progress',
+                    type: 'area',
+                    yAxis: 1,
+                    data: accumulative_data
+                }]
+            });
+        }
+    );
+});
+</script>
+<script>
+$(function() {
+    $.get(
+        "${api_url}"+"badges/",
+        data={"username": "${ user.username | h }"},
+        function( data ) {
+            for (i in data) {
+                $('#container-badges').append(
+                    '<img src="'+data[0].url+'" alt="Smiley face" height="100" width="100">'
+                );
+            }
+        }
+    );
+});
+</script>
+
   % if settings.FEATURES.get('ENABLE_DASHBOARD_SEARCH'):
     <%static:require_module module_name="js/search/dashboard/dashboard_search_factory" class_name="DashboardSearchFactory">
         DashboardSearchFactory();
@@ -127,7 +297,7 @@ if(localStorage["honor_code_skipped"] == undefined || localStorage["honor_code_s
 }
 .course-item .wrapper-action-more {
     position: absolute !important;
-    right: 5px; 
+    right: 5px;
     top: 155px;
     border: solid 1px #ccc;
     box-shadow: 1px 1px 5px #ccc;
@@ -157,7 +327,7 @@ if(localStorage["honor_code_skipped"] == undefined || localStorage["honor_code_s
     position: absolute;
 }
 .details .wrapper-course-image .cover{
-    border:none !important;    
+    border:none !important;
     float: none !important;
 }
 .dashboard {
@@ -246,6 +416,14 @@ input[data-bind-element]{
     %endif
 </div>
 
+<div class="chart-holder" style="display:flex; max-width:1180px; margin:0 auto; padding: 20px 0 0;">
+    <script src="https://code.highcharts.com/highcharts.js"></script>
+    <script src="https://code.highcharts.com/modules/exporting.js"></script>
+    <div id="container-chart" style="min-width: 310px; height: 400px; max-width: 600px; margin: 0 auto"></div>
+    <div id="container-progress" style="min-width: 310px; height: 400px; margin: 0 auto"></div>
+    <div id="container-badges" style="min-width: 310px; height: 400px; margin: 0 auto"></div>
+</div>
+
 <section class="container dashboard" id="dashboard-main">
       <section class="my-courses" id="my-courses" role="main" aria-label="Content">
       <div class="tab-container">
@@ -281,7 +459,7 @@ input[data-bind-element]{
         <% course_requirements = courses_requirements_not_met.get(enrollment.course_id) %>
         <% course_program_info = course_programs.get(unicode(enrollment.course_id)) %>
         <%include file = 'dashboard/_dashboard_course_listing.html' args="course_overview=enrollment.course_overview, enrollment=enrollment, show_courseware_link=show_courseware_link, cert_status=cert_status, can_unenroll=can_unenroll, credit_status=credit_status, show_email_settings=show_email_settings, course_mode_info=course_mode_info, show_refund_option=show_refund_option, is_paid_course=is_paid_course, is_course_blocked=is_course_blocked, verification_status=course_verification_status, course_requirements=course_requirements, dashboard_index=dashboard_index, share_settings=share_settings, user=user, course_program_info=course_program_info" />
-    % endfor 
+    % endfor
       </ul>
 
       <ul class="listing-courses" id="courses-completed">
@@ -365,6 +543,7 @@ input[data-bind-element]{
           </button>
         </div>
       </form>
+      <br>
     </div>
   % endif
 
-- 
2.5.0

